aws/iam_service_linked_role_test.go

```diff
@@ -0,0 +1,132 @@
+package aws
+
+import (
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	awsgo "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/gruntwork-io/cloud-nuke/config"
+	"github.com/gruntwork-io/cloud-nuke/util"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestListIamServiceLinkedRoles(t *testing.T) {
+	t.Parallel()
+
+	region, err := getRandomRegion()
+	require.NoError(t, err)
+
+	session, err := session.NewSession(&awsgo.Config{
+		Region: awsgo.String(region)},
+	)
+	require.NoError(t, err)
+
+	roleNames, err := getAllIamServiceLinkedRoles(session, time.Now(), config.Config{})
+	require.NoError(t, err)
+
+	assert.NotEmpty(t, roleNames)
+}
+
+func createTestServiceLinkedRole(t *testing.T, session *session.Session, name, awsServiceName string) error {
+	svc := iam.New(session)
+
+	input := &iam.CreateServiceLinkedRoleInput{
+		AWSServiceName: aws.String(awsServiceName),
+		Description:    aws.String("cloud-nuke-test"),
+		CustomSuffix:   aws.String(name),
+	}
+
+	_, err := svc.CreateServiceLinkedRole(input)
+	require.NoError(t, err)
+
+	return nil
+}
+
+func TestCreateIamServiceLinkedRole(t *testing.T) {
+	t.Parallel()
+
+	region, err := getRandomRegion()
+	require.NoError(t, err)
+
+	session, err := session.NewSession(&awsgo.Config{
+		Region: awsgo.String(region)},
+	)
+	require.NoError(t, err)
+
+	name := "cloud-nuke-test-" + util.UniqueID()
+	awsServiceName := "autoscaling.amazonaws.com"
+	iamServiceLinkedRoleName := "AWSServiceRoleForAutoScaling_" + name
+	roleNames, err := getAllIamServiceLinkedRoles(session, time.Now(), config.Config{})
+	require.NoError(t, err)
+	assert.NotContains(t, awsgo.StringValueSlice(roleNames), name)
+
+	err = createTestServiceLinkedRole(t, session, name, awsServiceName)
+	require.NoError(t, err)
+
+	roleNames, err = getAllIamServiceLinkedRoles(session, time.Now(), config.Config{})
+	require.NoError(t, err)
+	//AWSServiceRoleForAutoScaling_cloud-nuke-test
+	assert.Contains(t, awsgo.StringValueSlice(roleNames), iamServiceLinkedRoleName)
+}
+
+func TestNukeIamServiceLinkedRoles(t *testing.T) {
+	t.Parallel()
+
+	region, err := getRandomRegion()
+	require.NoError(t, err)
+
+	session, err := session.NewSession(&awsgo.Config{
+		Region: awsgo.String(region)},
+	)
+	require.NoError(t, err)
+
+	name := "cloud-nuke-test-" + util.UniqueID()
+	awsServiceName := "autoscaling.amazonaws.com"
+	iamServiceLinkedRoleName := "AWSServiceRoleForAutoScaling_" + name
+
+	err = createTestServiceLinkedRole(t, session, name, awsServiceName)
+	require.NoError(t, err)
+
+	err = nukeAllIamServiceLinkedRoles(session, []*string{&iamServiceLinkedRoleName})
+	require.NoError(t, err)
```

Comment:
Should we add an assertion here that the role is not in the result of `getAllIamServiceLinkedRoles`? I think that validates that the `nuke` was performed to our expectations, beyond not throwing an error. WDYT?

Reply:
Just to make sure I understand, are you suggesting creating a role that should not get deleted, and verify that it does indeed, not get deleted?

Reply:
Oh, I think I see what you're saying. Re-run getAllIamServiceLinkedRoles and make sure that the list doesn't contain the role any more.

Reply:
exactly!

Something like...

```go
roleNames, err = getAllIamServiceLinkedRoles(session, time.Now(), config.Config{})
require.NoError(t, err)

//AWSServiceRoleForAutoScaling_cloud-nuke-test
assert.NotContains(t, awsgo.StringValueSlice(roleNames), iamServiceLinkedRoleName)
```

Reply:
Addressed in 8572f99